File: Python/programmers/kakao/level4/블록_게임.py
def solution(board):
    n = len(board)
    answer = 0
    cnt = 0
    black = []
    
    while cnt <= 2:
        cnt += 1
        for col in range(n):
            idx = 0
            while True:
                if board[0][col] != 0:
                    break
                if idx >= n-1:
                    board[idx][col] = -1
                    black.append([idx, col])
                    break
                if board[idx+1][col] != 0:
                    board[idx][col] = -1
                    black.append([idx, col])
                    break
                idx += 1
        
        for row in range(n-1):
            for col in range(n-2):
                block = []
                if board[row][col] != 0:
                    for p in range(row, row+2):
                        for q in range(col, col+3):
                            block.append(board[p][q])
                if block.count(-1) == 2 and block.count(0) == 0 and len(set(block)) == 2:
                    answer += 1
                    for p in range(row, row+2):
                        for q in range(col, col+3):
                            board[p][q] = 0
                    for idx in black:
                        x, y = idx
                        board[x][y] = 0
                    black = []
                    cnt = 0
        
        for row in range(n-2):
            for col in range(n-1):
                block = []
                if board[row][col] != 0:
                    for p in range(row, row+3):
                        for q in range(col, col+2):
                            block.append(board[p][q])
                if block.count(-1) == 2 and block.count(0) == 0 and len(set(block)) == 2:
                    answer += 1
                    for p in range(row, row+3):
                        for q in range(col, col+2):
                            board[p][q] = 0
                    for idx in black:
                        x, y = idx
                        board[x][y] = 0
                    black = []
                    cnt = 0
    
    return answer


board = [[0,0,0,0,0,0,0,0,0,0],[0,0,0,0,0,0,0,0,0,0],[0,0,0,0,0,0,0,0,0,0],[0,0,0,0,0,0,0,0,0,0],[0,0,0,0,0,0,4,0,0,0],[0,0,0,0,0,4,4,0,0,0],[0,0,0,0,3,0,4,0,0,0],[0,0,0,2,3,0,0,0,5,5],[1,2,2,2,3,3,0,0,0,5],[1,1,1,0,0,0,0,0,0,5]]
print(solution(board)) # 2

File: Java/programmers/kakao/level4/블록게임.java
package CodeTest.Java.programmers.kakao.level4;

import java.util.*;

public class 블록게임 {
    public int solution(int[][] board) {
        int n = board.length; 
        int answer = 0;
        int cnt = 0;
        List<int[]> black = new ArrayList<>();

        while (cnt <= 2) {
            cnt++;
            for (int col=0; col<n; col++) {
                int idx = 0;
                while (true) {
                    if (board[0][col] != 0) break;
                    if (idx >= n-1) {
                        board[idx][col] = -1;
                        black.add(new int[]{idx, col});
                        break;
                    }
                    if (board[idx+1][col] != 0) {
                        board[idx][col] = -1;
                        black.add(new int[]{idx, col});
                        break;
                    }
                    idx++;
                }
            }

            for (int row=0; row<n-1; row++) {
                for (int col=0; col<n-2; col++) {
                    int minus = 0, zero = 0;
                    Set<Integer> block = new HashSet<>();
                    if (board[row][col] != 0) {
                        for (int p=row; p<=row+1; p++) {
                            for (int q=col; q<=col+2; q++) {
                                block.add(board[p][q]);
                                if (board[p][q] == -1) minus++;
                                if (board[p][q] == 0) zero++;
                            }
                        }
                    }
                    if (minus == 2 && zero == 0 && block.size() == 2) {
                        answer++;
                        for (int p=row; p<=row+1; p++) {
                            for (int q=col; q<=col+2; q++) {
                                board[p][q] = 0;
                            }
                        }
                        for (int[] idx : black) {
                            board[idx[0]][idx[1]] = 0;
                        }
                        black.clear();
                        cnt = 0;
                    }
                }
            }

            for (int row=0; row<n-2; row++) {
                for (int col=0; col<n-1; col++) {
                    int minus = 0, zero = 0;
                    Set<Integer> block = new HashSet<>();
                    if (board[row][col] != 0) {
                        for (int p=row; p<=row+2; p++) {
                            for (int q=col; q<=col+1; q++) {
                                block.add(board[p][q]);
                                if (board[p][q] == -1) minus++;
                                if (board[p][q] == 0) zero++;
                            }
                        }
                    }
                    if (minus == 2 && zero == 0 && block.size() == 2) {
                        answer++;
                        for (int p=row; p<=row+2; p++) {
                            for (int q=col; q<=col+1; q++) {
                                board[p][q] = 0;
                            }
                        }
                        for (int[] idx : black) {
                            board[idx[0]][idx[1]] = 0;
                        }
                        black.clear();
                        cnt = 0;
                    }
                }
            }
        }

        return answer;
    }
}
